Hardware wallet signer fixes #139

stdevAlDen opened the pull request:
Fixes #139

**Changes:**
- Remove `vendor/github.com/SkycoinProject/skycoin` git submodule, use it from vendors.
- Run [skywallet](https://github.com/fibercrypto/skywallet-mcu) in travis to run integration tests against it.
- Fix bug in `GetInputs`, previous cached value.
- Add [skywallet-go](github.com/fibercrypto/skywallet-go) dependency
- Implement signer iterator interface.
- GetSignerUID and GetSignerDescription could return error.
- Implement `EnumerateSignServices` for local/remote wallet.
- Implement get wallet matching the first address
- Add `TestTransactionSignInputFromDevice` for the hardware wallet signer.
- Implement signer for hardware wallet.
- Add `QSigner` and `SignerModel` models
- Add gui to select signer in simple mode for sending

**Does this change need to mentioned in CHANGELOG.md?**
yes

**Requires changes in protobuf specs?**
no

**Requires testing**
yes

Commit:
vendor packages ref #139

Commit:
Merge remote-tracking branch 'origin/olemis_t138_txn_sign' into stdevAlDen_t139_hardware_wallet_signer ref #139

Commit:
ignore github.com/therecipe/qt ref #139

Commit:
add hardware wallet singner implementation ref #139

Commit:
make device injectable and fix error handling in GetSignerUID ref #139

Commit:
generate mocks from Makefile ref #139

Commit:
Add TestGetSignerUIDShouldBeOk ref #139

Commit:
Add TestGetSignerUIDShouldFailForUninitializedDevice ref #139

Commit:
Add TestGetSignerUIDShouldFailOnDeviceError ref #139

Commit:
Add TestGetSignerUIDShouldFailForFailResponse ref #139

Commit:
Add TestGetSignerUIDShouldFailForInvalidMessageType ref #139

Commit:
fix import in tests ref #139

Commit:
Fix GetSignerDescription, error handling, urnPrefix and i18n ref #139

Commit:
Add TestGetSignerDescriptionShouldBeOk ref #139

Commit:
Add TestGetSignerDescriptionShouldFailForUninitializedDevice ref #139

Commit:
Add TestGetSignerDescriptionShouldFailOnDeviceError ref #139

Commit:
Add TestGetSignerDescriptionShouldFailForFailResponse ref #139

Commit:
Add TestGetSignerDescriptionShouldFailForInvalidMessageType ref #139

Commit:
Take some notes for i18n in GetSignerUID ref #139

Commit:
Run hardware wallet emulator on travis ref #139

Commit:
install protobf in travis ref #139

Commit:
install pip in travis ref #139

Commit:
install python protobuf in python ref #139

Commit:
install protobuf from pip in travis ref #139

Commit:
Revert "install protobuf from pip in travis ref #139"

This reverts commit 533b214454602dcbfa43fa547f848bb7698a79ff.

ref #139

Commit:
install protobuf version 2.6.1 ref #139

Commit:
Fix target to protobuf installation ref #139

Commit:
use python3 to install proto deps and track mocks ref #139

Commit:
Merge branch 'olemis_t138_txn_sign' into stdevAlDen_t139_hardware_wallet_signer

Conflicts:
	.travis.yml
	Makefile
	vendor/github.com/skycoin/hardware-wallet-go/src/skywallet/messages.go
	vendor/github.com/skycoin/hardware-wallet-go/src/skywallet/skywallet.go
	vendor/github.com/skycoin/hardware-wallet-protob/go/google/protobuf/descriptor.pb.go
	vendor/github.com/skycoin/hardware-wallet-protob/go/messages.pb.go
	vendor/github.com/skycoin/hardware-wallet-protob/go/types.pb.go
	vendor/golang.org/x/sys/windows/syscall_windows.go
	vendor/golang.org/x/sys/windows/zsyscall_windows.go
ref #139

Commit:
copy olemis code for integration tests ref #139

Commit:
use hardware wallet as a signer in integration tests ref #139

Commit:
Merge branch 'olemis_t138_txn_sign' into stdevAlDen_t139_hardware_wallet_signer ref #139

Commit:
Merge remote-tracking branch 'origin/olemis_t138_txn_sign' into stdevAlDen_t139_hardware_wallet_signer

Conflicts:
	.travis.yml
ref #139

Commit:
Merge branch 'olemis_t138_txn_sign' into stdevAlDen_t139_hardware_wallet_signer ref #139

Commit:
Merge branch 'olemis_t138_txn_sign' into stdevAlDen_t139_hardware_wallet_signer ref #139

Commit:
update integration tests from olemis ref #139

Commit:
Merge branch 'olemis_t138_txn_sign' into stdevAlDen_t139_hardware_wallet_signer

Conflicts:
	Gopkg.lock
	Gopkg.toml
	vendor/github.com/skycoin/hardware-wallet-go/src/skywallet/helper.go
	vendor/github.com/skycoin/hardware-wallet-go/src/skywallet/messages.go
	vendor/github.com/skycoin/hardware-wallet-go/src/skywallet/skywallet.go
	vendor/github.com/skycoin/hardware-wallet-protob/go/google/protobuf/descriptor.pb.go
	vendor/github.com/skycoin/hardware-wallet-protob/go/messages.pb.go
	vendor/github.com/skycoin/hardware-wallet-protob/go/types.pb.go
ref #139

Commit:
Merge branch 'olemis_t138_txn_sign' into stdevAlDen_t139_hardware_wallet_signer ref #139

Commit:
create some util function for device interaction in tests ref #139

Commit:
Hardware walt integration tests:
 - Create a setup function for hw tests(RegisterAltcoin, SetMnemonic, ...).
 - Make testTransactionSignInput test a resuable function.
 - Make testTransactionSignInput resued from TestTransactionSignInput and TestTransactionSignInputFromDevice
 - Set a callback to sign transactions
 - Fix SignTransaction to check for IsFullySigned.

ref #139

Commit:
Merge remote-tracking branch 'origin/olemis_t138_txn_sign' into stdevAlDen_t139_hardware_wallet_signer

Conflicts:
	.travis.yml
	src/coin/skycoin/models/wallet_test.go
ref #139

Commit:
Merge remote-tracking branch 'upstream/develop' into stdevAlDen_t139_hardware_wallet_signer ref #139

Commit:
find wallet from first addr

- Add `GetWallet(firstAddr string) (Wallet, error)` to `core.WalletEnv` interface.
- Add `GetWallet` implementation to `WalletDirectory` and `WalletNode`

ref #139

Commit:
refactor getWallet into a single reusable function ref #139

Commit:
add function to get first address from hardware wallet ref #139

Commit:
check periodically for hardware wallet ref #139

Commit:
introdue a hardware wallet repesentation in the model
 - Introducing a new field to the qwallet model: HasHardwareWallet
 - Make sniffHw a single shot function, goroutina looks as not working with this framework
 - updateWallet find and notify about change in the relevant row of the model
 - primitive handling for connection to hardware wallet was lose
 - fix AddressGen from hw for the first address, wrong start index

ref #139

Commit:
Introduce a dumy representation of hw signer and periodically checking ref #139

Commit:
Fix getWallet from address to work across full iterators
ref #139

Commit:
Fix sniffHw, handle error if can not find a wallet matching the hardware one
ref #139

Commit:
Fix endRanger index in sniffHw ref #139

Commit:
Track the row where the signer is connected on. ref #139

Commit:
Merge remote-tracking branch 'upstream/develop' into stdevAlDen_t139_hardware_wallet_signer ref #139

Commit:
Merge remote-tracking branch 'upstream/develop' into stdevAlDen_t139_hardware_wallet_signer ref #139

Commit:
Merge remote-tracking branch 'upstream/develop' into stdevAlDen_t139_hardware_wallet_signer

Conflicts:
	Makefile
ref #139

Commit:
Remove test forlder at root project dir, integrations tests will be inside the same packages udner test ref #139

Commit:
Refactor sniffHw to more reusable functions (sniffHw, attachHwAsSigner and hwMatchWallet) ref #139

Commit:
implement signer iterator for EnumerateSignServices ref #139

Commit:
fix autogenerated moc for hw devicer ref #139

Commit:
wallet enumerate signers enumerate it selft as signer ref #139

Commit:
fix bug in get wallet from address ref #139

Commit:
Impllement calback fro sign transaction interaction with device ref #139

Commit:
Add gui for sign transaction using hardware wallet as signer ref #139

Commit:
update vendor folder ref #139

Commit:
fix import path for skywallet-go and skywallet-protob ref #139

Commit:
Merge remote-tracking branch 'upstream/develop' into stdevAlDen_t139_hardware_wallet_signer

Conflicts:
	.gitignore
	src/coin/skycoin/models/wallet_test.go
	src/models/walletsModel.go
	src/ui/SubPageSendSimple.qml
ref #139

Commit:
update vendore ref #139

Commit:
fix deps rule in makefile ref #139

Commit:
udate vendors ref #139

Commit:
add signers model ref #139

Commit:
add ci scripts ref #139

Commit:
fix travis yml format ref #139

Commit:
add testdata for skycoin in vendor folder ref #139

Commit:
Merge remote-tracking branch 'simelo/olemis_t234_global_signers' into stdevAlDen_t139_hardware_wallet_signer

Conflicts:
	src/coin/skycoin/models/wallet.go
	src/coin/skycoin/models/wallet_test.go

Commit:
Merge remote-tracking branch 'simelo/olemis_t234_global_signers' into stdevAlDen_t139_hardware_wallet_signer

Commit:
integrate changes(about signers) in implementation ref #139

Commit:
Merge remote-tracking branch 'simelo/olemis_t234_global_signers' into stdevAlDen_t139_hardware_wallet_signer

Conflicts:
	Makefile
ref #139

Commit:
fix signerts iterator for DefaultTxnSignerIterator implementation ref #139

Commit:
Add ReadyForTxn implementation for hardware wallet ref #139

Commit:
attachHwAsSigner register the signer in a global altCoinManager ref #139

Commit:
filter signers by wallet using signer.ReadyForTxn ref #139

Commit:
fix rule for mocks generation in makefile ref #139

Commit:
update vendors ref #139

Commit:
fix unit test for hardware wallet ref #139

Commit:
update mocks ref #139

Commit:
change iterator convention according to 1b00d068e1f82587b563fc809f0cc3b9a380869e ref #139

Commit:
fix tests ref #139

Commit:
Merge remote-tracking branch 'upstream/develop' into stdevAlDen_t139_hardware_wallet_signer ref #139

Commit:
fix test rules in makefile ref #139

Commit:
fix gen-mocs rules ref #139

Commit:
fix bug introduced by a conflict resolution ref #139

Commit:
disable addrIndex field in sign transction using hardware wallet ref #139

Commit:
Merge remote-tracking branch 'upstream/develop' into stdevAlDen_t139_hardware_wallet_signer

Conflicts:
	.travis.yml
ref #139

Commit:
fix linter error ref #139

Commit:
Merge remote-tracking branch 'upstream/develop' into stdevAlDen_t139_hardware_wallet_signer ref #139

Commit:
AddSignature(index uint64, signature []byte) error for Transaction interface ref #139

Commit:
Add a clonable interface
make core.Transaction a clonable object
ref #139

Commit:
fix copyTransaction, preserve original slice capacity ref #139

Commit:
fix GetInputs for SkycoinUninjectedTransaction ref #139

Commit:
fix GetOutputs for SkycoinUninjectedTransaction ref #139

Commit:
fix Clone, handle null input ref #139

Commit:
add StrSlice2IntSlice ref #139

Commit:
Fix sign transaction with hardware wallet:
 - Refactor `SignTransaction`, make `getInputs` and `getOutputs` resuable methods.
 - Create a transaction deep copy before signing.
 - Assert `msg.Kind` after signing the transaction.
 - Use specific indexes from function argments to sign secific outputs
ref #139

Commit:
track a mock object ref #139

Commit:
improve device instance reutilization ref #139

Commit:
improve error handling and msg reporting it ref #139

Commit:
refactor GetSignerUID and GetSignerDescription making use of getDeviceFeatures ref #139

Commit:
Merge remote-tracking branch 'upstream/develop' into stdevAlDen_t139_hardware_wallet_signer

Conflicts:
	.travis.yml
ref #139

Commit:
change GetSignerUID and GetSignerDescription from TxnSigner to consider some possible errors ref #139

Commit:
make trigger gen-mocks on test* rules in makefile ref #139

Commit:
if no device found ReadyForTxn in SkyWallet should inform about this at Debug level only ref #139

Commit:
fix error name in response ref #139

Commit:
handle error in signer uid and description in signers model ref #139

Commit:
getAllIndexesFromTxn if not indexes specified in sign transaction ref #139

Commit:
Merge remote-tracking branch 'upstream/develop' into stdevAlDen_t139_hardware_wallet_signer

Conflicts:
	.travis.yml
	Gopkg.lock
	Gopkg.toml
	src/coin/skycoin/models/coin.go
	src/coin/skycoin/models/wallet.go
	src/coin/skycoin/models/wallet_test.go
	src/core/main.go
	src/errors/error.go
	src/models/walletsModel.go
	src/ui/PageSend.qml
	src/ui/SubPageSendSimple.qml
	src/util/util.go
ref #139

Commit:
Merge remote-tracking branch 'upstream/develop' into stdevAlDen_t139_hardware_wallet_signer

Conflicts:
	.travis.yml
	Makefile
	src/models/walletsModel.go
	src/util/coin.go
ref #139

Commit:
change project name to lowercase from FiberCryptoWallet to fibercryptowallet ref #139

Commit:
remove AddSignature from transaction ref #139

Commit:
make TestTransactionSignInput a reusable function to run it from hardware wallet ref #139

Commit:
fix signing transactio with device

move from generic implementatio to specific type assertions(readableTransactio and rawTransaction) and update header
ref #139

Commit:
Merge remote-tracking branch 'upstream/develop' into stdevAlDen_t139_hardware_wallet_signer ref #139

Commit:
run xvfb as service ref #139

olemis commented on .travis.yml:
```diff
@@ -36,14 +41,33 @@ matrix:
     - make install-linters
     - make install-coveralls
+    before_script:
+    # Install hardware wallet for running tests against emulator
```
could you please implement this as `make` target or CI script ?

olemis commented on .travis.yml:
```diff
@@ -36,14 +41,33 @@ matrix:
     script:
     - make clean
     - DEFAULT_TARGET=linux make build-docker
+    - PYTHON=python3 PIP=pip3 make -C tmp/hardware-wallet clean
```
move to `before_script`

Commit:
tests no need to depend on gen-mocks as we are tracking the mock files ref #139

olemis reviewed (changes requested):
Move `src/hardware` => `src/contrib/skywallet`

olemis commented on Makefile:
```diff
@@ -231,9 +231,14 @@ clean-build: ## Remove temporary files
 	mockery -all -output src/coin/mocks -outpkg mocks -dir src/core
 	find src/coin/mocks/ -name '*.go' -type f -print0
 
+test-hw: ## Run Hardware wallet tests
```
Rename `test-hw` => `test-skyhw`

olemis commented on ci-scripts/integration-test.sh:
```diff
@@ -0,0 +1,50 @@
+#!/bin/bash
```
What's this file for ?

olemis commented on src/coin/skycoin/models/coin.go:
```diff
@@ -1,7 +1,9 @@
 import (
+	stdErr "errors"
 	"fmt"
+	"github.com/sirupsen/logrus"
```
Use logger instance . Ask @stdevMac if in doubt .

olemis commented on src/coin/skycoin/models/cipher.go:
```diff
@@ -47,6 +47,17 @@ func (addr *SkycoinAddress) GetCryptoAccount() core.CryptoAccount {
 	return addr
 }
 
+func (addr *SkycoinAddress) Clone() (interface{}, error) {
```
Not sure about object cloning methods . Are those still needed?

olemis commented on src/coin/skycoin/models/signer_iterator.go:
```diff
@@ -0,0 +1,32 @@
+package skycoin
```
Duplicated functionality ? Same as `DefaultTxnSignerIterator` ?

olemis commented on src/coin/skycoin/models/sky_test.go:
```diff
@@ -1,188 +0,0 @@
-package skycoin
```
Functions in this file have to remain separate from he rest , since they've been copied verbatim from Skycoin code base .

olemis commented on src/coin/skycoin/models/skyapi_mock_test.go:
```diff
@@ -1,108 +0,0 @@
-package skycoin
```
Functions in this file have to remain apart of the rest due to their specific purpose of mocking real classes .

olemis commented on src/coin/skycoin/models/testsutil.go:
```diff
@@ -0,0 +1,483 @@
+	"path/filepath"
+	"strings"
+	"testing"
+	//"crypto/rand"
```
Remove commented package imports ?

olemis commented on src/coin/skycoin/models/testsutil.go:
```diff
@@ -0,0 +1,483 @@
+package skycoin
```
Rename `testsutil.go` => `testutils.go`

olemis commented on src/coin/skycoin/models/testsutil.go:
```diff
@@ -0,0 +1,483 @@
+	return
+}
+
+func randBytes(t *testing.T, n int) []byte {
```
This is not specific to Skycoin altcoin plugin . Please move onto `util` package.

olemis commented on src/coin/skycoin/models/wallet.go:
```diff
@@ -279,6 +279,10 @@ type WalletNode struct {
 	poolSection string
 }
 
+func (wltEnv *WalletNode) GetWallet(firstAddr string) (core.Wallet, error) {
```
is this tested ?

olemis commented on src/coin/skycoin/models/wallet.go:
```diff
@@ -1121,6 +1145,7 @@ func (wlt *LocalWallet) signSkycoinTxn(txn core.Transaction, pwd core.PasswordRe
 		// Raw transaction
 		unTxn, ok := txn.(*SkycoinUninjectedTransaction)
 		if !ok {
+			// FIXME what error?
```
What's the final word on this ?

olemis commented on src/core/coin.go:
```diff
@@ -15,8 +15,15 @@ const (
 )
 
+// Define a type that can perform a deep copy from the original object
+type Clonable interface {
```
Not sure this still needed . Remove ?

olemis commented on src/core/coin.go:
```diff
@@ -15,8 +15,15 @@ const (
+
 // Transaction encapsulates the contract for atomic transfers of coins
 type Transaction interface {
+	Clonable
```
Cloning capabilities are not requirements of transactions.

olemis commented on src/core/coin.go:
```diff
@@ -51,6 +58,7 @@ type TransactionIterator interface {
 
 // TransactionInput provides cryptographic proof of spending funds
 type TransactionInput interface {
+	Clonable
```
Cloning capabilities are not requirements of transactions.

olemis commented on src/core/coin.go:
```diff
@@ -74,6 +82,7 @@ type TransactionInputIterator interface {
 
 // TransactionOutput provides cryptographic proof of funds transfer
 type TransactionOutput interface {
+	Clonable
```
Cloning capabilities are not requirements of transactions.

Commit:
run hardware wallet tests for linux and osx only ref #139

olemis commented on src/core/wallet.go:
```diff
@@ -79,10 +79,13 @@ type SeedGenerator interface {
 	GetWalletSet() WalletSet
+	// GetWallet allow you to get a wallet from the first address in the
+	// deterministic sequence, error if not found or some thing was happen
+	GetWallet(firstAddr string) (Wallet, error)
```
Rename `GetWallet()` => `LookupWallet()`

olemis commented on src/errors/error.go:
```diff
@@ -49,4 +49,17 @@ var (
 	// ErrInvalidValue invalid value was supplied in to function
 	ErrInvalidValue = errors.New("Value errors")
+	// ErrInvalidTypeAssertion the current value have not the expected type
+	ErrInvalidTypeAssertion = errors.New("invalid type assertion")
```
Start error message with capital letters.

olemis commented on src/hardware/sky-wallet.go:
```diff
@@ -0,0 +1,446 @@
+}
+
+const (
+	urnPrefix = "Hardware:SkyWallet:"
```
Shouldn't this be `signer:skywallet` ?

olemis commented on src/hardware/sky-wallet.go:
```diff
@@ -0,0 +1,446 @@
+	"github.com/fibercrypto/skywallet-go/src/skywallet/wire"
+	"github.com/fibercrypto/skywallet-protob/go"
+	"github.com/gogo/protobuf/proto"
+	"github.com/sirupsen/logrus"
```
Do not use logrus directly , but loggers . Ask @stdevMac if in doubt .

olemis commented on src/hardware/sky-wallet.go:
```diff
@@ -0,0 +1,446 @@
+func HwFirstAddr(dev skyWallet.Devicer) (string, error) {
+	msg, err := dev.AddressGen(1, 0, false, skyWallet.WalletTypeDeterministic)
+	if err != nil {
+		// TODO i18n
```
No translation for log messages.

olemis commented on src/hardware/sky-wallet.go:
```diff
@@ -0,0 +1,446 @@
+		transactionOutput.Address = proto.String(output.Address.String())
+		transactionOutput.Coin = proto.Uint64(output.Coins)
+		transactionOutput.Hour = proto.Uint64(output.Hours)
+		// FIXME: should be possible to send the address index
```
What about these lines commented out ?

olemis commented on src/hardware/sky-wallet.go:
```diff
@@ -0,0 +1,446 @@
+	return skyTxn, nil
+}
+
+func rawTxn2Transaction (txn *skycoin.SkycoinUninjectedTransaction) (*coin.Transaction, error) {
```
Expand argument type scope . All instances of `SkycoinTxn` implement `EncodeSkycoinTransaction()`

Commit:
move some instalation rules from travis.yml to ci-scripts ref #139

Commit:
refactor rename test-hw ruule to test-skyhw ref #139

Commit:
run sky-hw emulator in before_script for linux and osx ref #139

Commit:
0 len slice cover the nil case ref #139

Commit:
making the transaction a clonable object is not a requeriment ref #139

Commit:
use fibercrypto logger wrapper ref #139

Commit:
fix urnPrefix for skyWallet ref #139

Commit:
add unit tests for LookupWallet ref #139

Commit:
All instances of SkycoinTxn implement EncodeSkycoinTransaction ref #139

Commit:
keep mocks for test in a separated file (skyapi_test_mocks.go) ref #139

Commit:
keep util functions for tests in a separate file (sky_test_utils.go) ref #139

Commit:
renamed:    src/coin/skycoin/models/testsutil.go -> src/coin/skycoin/models/testsutils.go ref #139

Commit:
renamed:    src/hardware/ -> src/contrib/skywallet/ ref #139

Commit:
Merge remote-tracking branch 'simelo/stdevHsequeda_t231_address_verify' into stdevAlDen_t139_hardware_wallet_signer

Conflicts:
	src/coin/skycoin/models/wallet_test.go
	src/util/util.go
ref #139

Commit:
be able to sign bip44 transactions with skywallet device ref #139

Commit:
no need for confirmation in the device for change address ref #139

Commit:
Merge remote-tracking branch 'upstream/develop' into stdevAlDen_t139_hardware_wallet_signer

Conflicts:
	src/coin/skycoin/models/cipher.go
	src/coin/skycoin/models/cipher_test.go
	src/coin/skycoin/models/coin.go
	src/coin/skycoin/models/main.go
	src/coin/skycoin/models/wallet.go
	src/coin/skycoin/models/wallet_test.go
	src/core/main.go
	src/util/cipher.go
	src/util/util.go
ref #139

Commit:
use a more generic parser for indices in SignTransaction ref #139

Commit:
fix travisyml file format ref #139

Commit:
fix BINPATH for linux and windows in makefile ref #139

Commit:
update mocks ref #139

Commit:
fix layout for signer selector ref #139

Commit:
declare pwd as core.PasswordReader to avoid wrong type inference ref #278

Commit:
Merge remote-tracking branch 'upstream/develop' into stdevAlDen_t139_hardware_wallet_signer

Conflicts:
	src/models/walletsManager.go
ref #139

olemis committed:
Merge remote-tracking branch 'fc/develop' into stdevAlDen_t139_hardware_wallet_signer

Commit:
Merge branch 'stdevAlDen_t139_hardware_wallet_signer' of github.com:simelo/FiberCryptoWallet into stdevAlDen_t139_hardware_wallet_signer

Conflicts:
	src/models/walletsManager.go
ref #139

Commit:
fix Gopkg.toml ref #139

Commit:
install Skycoin API mocks in testing code for sky-wallet ref #139

Commit:
handle multiwallet registration for signer ref #139

Commit:
let the hardware sniffer run in an async goroutine ref #139

Commit:
Merge remote-tracking branch 'simelo/stdevMauricio1802_t278_Unable_to_sign_transaction' into stdevAlDen_t139_hardware_wallet_signer ref #139

Commit:
force command to be atomic as a single sequence ref #139

a temporary patch related to https://github.com/fibercrypto/skywallet-go/issues/163

Commit:
update vendors to use the Sequencer implementation of Devicer from skywallet-go WIP ref #139

Commit:
integrate sequencer for device interaction ref #139

Commit:
fix tests with new Devicer implementation ref #139

olemis committed:
Merge remote-tracking branch 'fc/develop' into stdevAlDen_t139_hardware_wallet_signer

olemis committed:
Merge remote-tracking branch 'simelo_alt/stdevAlDen_t139_hardware_wallet_signer' into stdevAlDen_t139_hardware_wallet_signer

olemis commented:
@stdevAlDen does code need to be updated after merging fibercrypto/skywallet-go#164 ?

olemis committed:
Merge remote-tracking branch 'fc/develop' into stdevAlDen_t139_hardware_wallet_signer

Commit:
Merge remote-tracking branch 'upstream/develop' into stdevAlDen_t139_hardware_wallet_signer

Conflicts:
	Gopkg.lock
	Makefile
	src/coin/skycoin/models/main_test.go
	src/errors/error.go
ref #139

Commit:
Merge remote-tracking branch 'upstream/develop' into stdevAlDen_t139_hardware_wallet_signer

Conflicts:
	Makefile
ref #139

Commit:
fix LookupSignerByUID ref #139

Commit:
fix imports ref #139

Commit:
fix tests to use new GetSignerUID interface (able to return error) ref #139

Commit:
set the emulator ip addres as a test argument ref #139

Commit:
Update vendors ref #139

Commit:
shares the host's networking namespace with the docker container ref #139

Commit:
run test-skyhw inside the docker continer ref #139

Commit:
Merge remote-tracking branch 'upstream/develop' into stdevAlDen_t139_hardware_wallet_signer

Conflicts:
	src/ui/SubPageSendSimple.qml
ref #139

Commit:
fix emulator ip for integration tests ref #139

Commit:
do not remove golden files from skycoin ref #139

Commit:
Remove SkycoinProject as submodule

Commit:
add github.com/SkycoinProject/skycoin to vendors

Commit:
build before run hardware wallet tests ref #139

Commit:
no need to set the ip address as docker is using the host network ref #139

Commit:
no need for send button ack in tests as sequencer solve this internally #139

Commit:
docker not suported for osx images on travis ref #139

stdevMac commented on .gitignore:
```diff
@@ -2,6 +2,7 @@
 *.so
 
 # Distribution / packaging
+vendor/github.com/therecipe/qt/
```
@stdevAlDen I think we need this in the gitignore, cause some of the deps in therecipe/qt are needed and static, example `widgets`. Any special reason to remove it?

stdevAlDen replied:
> @stdevAlDen I think we need this in the gitignore, cause some of the deps in therecipe/qt are needed and static, example `widgets`. Any special reason to remove it?

@stdevMac when therecipe qt binding is inside the vendors the project does not build (at least for me), no other thing...

olemis replied:
but adding it in `.gitignore` hides the fact that there is something wrong with the working copy .

Commit:
generate coverage report from skywallet ref #139

Commit:
update vendors ref #139

Commit:
get signature as hex value from skycoin API ref #139

Commit:
fix find Addr Index From Input

address index for the spend output related for the input is used to find the priv/pub key pair
ref #139

Commit:
Merge remote-tracking branch 'upstream/develop' into stdevAlDen_t139_hardware_wallet_signer

Conflicts:
	Gopkg.lock
ref #139

Commit:
update vendors ref #139

Commit:
update changes log ref #139

Commit:
set the wallet as default signer in the view ref #139

Commit:
fix: be able to use the wallet it selft on having multiple signers ref #139

Commit:
fix tests: use a skycoin API client from a mock ref #139

Commit:
fix test: load more addresess to make getAddrIndex working when using GetLoadedAddresses ref #139

olemis commented on src/coin/skycoin/models/wallet.go:
```diff
@@ -812,7 +816,31 @@ func NewWalletDirectory(dirPath string) *WalletDirectory {
 	// Implements WallentEnv interface
 	WalletDir  string
-	wltService *SkycoinLocalWallet
+	wltService core.PersistibleSet
```
@stdevAlDen why this change ?

olemis commented on src/contrib/skywallet/sky-wallet.go:
```diff
@@ -0,0 +1,428 @@
+	"github.com/fibercrypto/fibercryptowallet/src/coin/skycoin/params"
+	"github.com/fibercrypto/fibercryptowallet/src/coin/skycoin/skytypes"
+	"github.com/fibercrypto/fibercryptowallet/src/core"
+	fce "github.com/fibercrypto/fibercryptowallet/src/errors"
```
`fce` => `fcwerrors`

olemis commented on src/contrib/skywallet/sky-wallet.go:
```diff
@@ -0,0 +1,428 @@
+	return inputs, nil
+}
+
+func spendingOutputFromRemote(inputHash cipher.SHA256) (*readable.SpentOutput, error) {
```
Duplicated . Reuse the logic in Skycoin altcoin plugin .

olemis commented on src/contrib/skywallet/sky-wallet.go:
```diff
@@ -0,0 +1,428 @@
+	return
+}
+
+func readableTxn2Transaction(txn skytypes.ReadableTxn) (*coin.Transaction, error) {
```
Duplicated . Reuse Skycoin altcoin plugin code .

olemis commented on src/contrib/skywallet/sky-wallet.go:
```diff
@@ -0,0 +1,428 @@
+
+func rawTxn2Transaction (txn skytypes.SkycoinTxn) (*coin.Transaction, error) {
+	buf, err := txn.EncodeSkycoinTransaction()
+	if err != nil {
```
What happens on error ?

olemis commented on src/contrib/skywallet/sky-wallet.go:
```diff
@@ -0,0 +1,428 @@
+	return skyTxn, nil
+}
+
+func rawTxn2Transaction (txn skytypes.SkycoinTxn) (*coin.Transaction, error) {
```
Duplicate . Reuse Skycoin altcoin plugin code.

olemis commented on src/contrib/skywallet/sky-wallet.go:
```diff
@@ -0,0 +1,428 @@
+	return &skyTxn, nil
+}
+
+func toTransaction(txn core.Transaction) (*coin.Transaction, error) {
```
Duplicate . Reuse Skycoin altcoin plugin code

olemis commented on src/contrib/skywallet/sky-wallet.go:
```diff
@@ -0,0 +1,428 @@
+	unTxn, err := skycoin.NewUninjectedTransaction(txn, fee)
+	if err != nil {
+		logSkyWallet.WithError(err).Errorln("unable to create uninjected transaction")
+		return nil, fce.ErrTxnSignFailure
```
Why sign failure if not signing transaction yet ?

olemis commented on src/contrib/skywallet/sky-wallet.go:
```diff
@@ -0,0 +1,428 @@
+	transactionInputs, err := getInputs(sw.wlt, *txn, idxs)
+	if err != nil {
+		logSkyWallet.WithError(err).Errorln("unable to get inputs")
+		return nil, fce.ErrTxnSignFailure
```
Use different error conditions in each case to identify failure path.

olemis replied:
Let me take a second try on it .

olemis commented on src/contrib/skywallet/sky-wallet.go:
```diff
@@ -0,0 +1,428 @@
+	features, err := sw.getDeviceFeatures()
+	if err != nil {
+		logSkyWallet.WithError(err).Error("unable to get device features")
+		return "", fce.ErrHwUnexpected
```
No error , but generic placeholder ...

olemis commented on src/contrib/skywallet/sky-wallet.go:
```diff
@@ -0,0 +1,428 @@
+package hardware
```
Package name should match folder name . `skywallet` ?

olemis commented on src/core/wallet.go:
```diff
@@ -24,6 +24,11 @@ type WalletSet interface {
 	SupportedWalletTypes() []string
 }
 
+type PersistibleSet interface {
```
Need be renamed .

olemis commented on src/models/walletsManager.go:
```diff
@@ -618,6 +618,10 @@ func (walletM *WalletManager) signTxn(wltIds, address []string, source string, t
 			logWalletManager.WithError(err).Warnf("No signer %s for wallet %v", source, wlts[0])
 			return nil
 		}
+		if suid, err := signer.GetSignerUID(); err != nil && wlts[0].GetId() == string(suid) {
```
condition should be `err == nil` , right ?

olemis commented on src/models/walletsModel.go:
```diff
@@ -113,6 +192,20 @@ func (walletModel *WalletModel) data(index *core.QModelIndex, role int) *core.QV
+				}
+				return false
+			}
+			// FIXME: consider a double checking here instead of hadHwConnected
```
What about this `FIXME` tag ? Shall this remain ?

olemis commented on test/integration/util/hardware-wallet.go:
```diff
@@ -0,0 +1,23 @@
+package util
```
Shouldn't this be under `skywallet` contrib package ?

olemis committed:
Merge remote-tracking branch 'fc/develop' into stdevAlDen_t139_hardware_wallet_signer

olemis committed:
Merge remote-tracking branch 'simelo_alt/stdevAlDen_t139_hardware_wallet_signer' into stdevAlDen_t139_hardware_wallet_signer

olemis committed:
[errors] refs #139 - Fix ErrNotImplemented redeclared

olemis committed:
[sky] refs #139 - Adjust test comparing params.DataRefreshTimeout vs SkycoinBlockchain.CacheTime

olemis committed:
[sky] refs #139 - Correct comparison between params.DataRefreshTimeout and SkycoinBlockchain.CacheTime

olemis committed:
[ci] refs #139 - Remove duplicate declaration of make test-skyhw target

olemis committed:
[ci] refs #139 - Increase Skycoin test suite timeout

olemis committed:
[ci] refs #139 - Do not make build prior to make test-skyhw

olemis committed:
[contrib] [models] refs #139 - Handle errors after changing signature of GetSpentOutput and GetAddress

... still WiP

Commit:
improve error handling in derivationType determination ref #139

Commit:
fix mocks generation rule ref #139

Commit:
fix number scale in tests ref #139

Commit:
fix rawTxn2Transaction, handle errors ref #139
